feat: Express Build mode (#2018)
vaadin/docs: articles/configuration/development-mode/index.asciidoc

---
title: Development Mode

> description: Vaadin Development Mode is aimed at making application development faster by enabling easier debugging with a live-reload and front-end dev server.

Reviewer: ```suggestion
description: Vaadin Development Mode makes application development faster by enabling easier debugging with a live-reload and front-end dev server.
```

order: 20
---

== Overview

> Vaadin provides a mode, called "Development Mode", that enables easier debugging and fast start-up.

Reviewer 2: Not certain I would call the startup fast if you need to get all the node things.
Perhaps introduce it more as `Vaadin development mode contains helpful debugging tools and log output`, as a view into what is different from production mode.

> Depending on how much front-end customisations does your project have, you may want to choose running the application with the front-end dev server or without it.

Reviewer 2: ```suggestion
When working with a frontend heavy project, there can be a frontend development server running for fast compilation when frontend files change.
```
`frontend`, `front end` or `front-end` what is our documentation using? So confusing.

Reviewer 2: This change should then affect the whole rest of this paragraph, perhaps drop most of it and just refer to the mode specific topics that should then cover what they are/do

- no `com.vaadin:vaadin-dev-bundle` dependency added to the project

Custom theme files are served separately from the front-end bundle, i.e. they can be modified on the fly and the browser live-reload would trigger refresh immediately, no re-compilation is needed.
When customising a theme for the application with no dev server used, the following points should be taken into account:

- Vaadin component's styles are set up in [filename]`styles.css` (or in the imported style sheet file located in the theme directory) with the `::part()` selector.
- Use the `assets` block in the [filename]`theme.json` file to specify the external assets, like fonts, images or style sheets, to be added into the "application" dev bundle, so that they can be visible and used in the application.

See <<{articles}/styling/custom-theme/custom-theme-configuration#,Theme Configuration>> for more information.

This mode is available since *Vaadin 24.0* behind a feature flag `com.vaadin.experimental.expressBuild`, see <<{articles}/configuration/feature-flags#,Feature Flags>> for more information about the feature flags.

== Front-End Dev Server For Easier Debugging

> Vaadin uses the front-end dev server provided by Vite for enabling the client-side live-reload, so that the JavaScript/TypeScript changes made are immediately shown in the browser. Vite is optimised to re-compile only what is actually changed, which speeds up the hot reload.

Reviewer 2: ```suggestion
Vaadin uses Vite to speed up frontend development by enabling the client-side live-reload, so that JavaScript/TypeScript changes made are immediately updated to the browser. Vite only re-compiles what has actually changed, making hot reload fast.
```

> Vite is distributed as an npm/pnpm package and, as a consequence, requires the Node.js and npm/pnpm installation.

Reviewer 2: ```suggestion
Vite installation is made through npm/pnpm and running Vite requires Node.js.
```

See <<{articles}/configuration/development-mode/node-js#,Node.js installation>> and <<{articles}/configuration/development-mode/npm-pnpm#,Configuring front-end package manager>> for more information.

> This mode should be used in case of a components/view made using <<{articles}/create-ui/templates#,Templates>> are present in the project.

Reviewer 2: ```suggestion
This mode is recommended if components/views use <<{articles}/create-ui/templates#,Templates>> and are edited often.
```
Just having lots of templates that are not touched in the project should not warrant using Vite over express mode application bundle.

Development mode provides a helpful debugging tools and information (the development tools window) available in the browser.

<<{articles}/production#,Production mode>> aims to send as little data as possible to the browser of the end user and otherwise also optimize the performance wherever possible.
All development tools and debug information are disabled.

Development mode can further be divided into two modes:

- Using a pre-compiled front-end bundle
- Running with a front-end development server.

The difference between these modes is in how the front-end files are handled: when the front-end development server is up and running, it serves JavaScript, CSS and other front-end files to the browser one by one when they are requested, instead of the Java server the application is running on.
Otherwise, the front-end files are pre-compiled into a front-end bundle and this bundle is served to the browser similarly as in the production mode.

[role="since:com.vaadin:vaadin@V24"]
== Pre-Compiled Front-End Bundle For Faster Start-Up

> The build and start-up time can be improved significantly if no front-end installation or re-compilation is done at any point: installing Node.js, running Vite and npm/pnpm packages installation.

Author: I propose to keep mentioning Node.js, Vite and npm/pnpm here to let reader know what is meant by "front-end installation or recompilation" and what exactly we don't install to get the benefit.

> The build and start-up time can be improved significantly if no front-end installation or re-compilation is done at any point: installing Node.js, running Vite and npm/pnpm packages installation.

Reviewer 2: Well it keeps node.js, vite and npm, just doesn't mention pnpm (as that's not the default anymore)
Perhaps then set it to 2 parts `This means not installing the tools Node.js, npm/pnpm and downloading npm packages, nor running the Vite dev server`

If your project only uses the standard Vaadin Flow components or third-party Vaadin add-ons (e.g. from Vaadin Directory), including only Flow components, Vaadin then skips the re-compilation and uses the default set of components, themes and front-end files shipped with the Vaadin platform.

On the other hand, once any custom JavaScript/TypeScript codes or add-ons

> with front-end customisations are added to the project, the re-compilation is triggered by Vaadin during next start-up using npm/pnpm and Vite.

Author: The first modified paragraph looks a bit broken with the extra "adding". Would it be fine:
```
Adding any custom JavaScript/TypeScript or add-ons with front-end customisations to the project triggers the front-end re-compilation with Vite during next application's start-up.
Any new or missing front-end packages are downloaded using npm/pnpm before build.
```

Vaadin does it automatically and is able to spot the front-end customisations.
Making a new front-end bundle takes time still, but then all the output files (located in the `dev-bundle` directory inside project's root) can be added to the Version Control System, so that other developers can fetch it and run the application as though it uses a default bundle.

Thus, Vaadin defines two origins of the front-end development bundle: the "default" bundle updated and provided by the Vaadin platform in each release version, and the "application" bundle, which is made by local run of the project and taking into account the specific front-end customisations.
To let the "default" bundle be visible for Vaadin, one should add the dependency `com.vaadin:vaadin-dev-bundle` (with no version, since it's controlled by Vaadin platform).

Vaadin makes a new "application" bundle in the following cases:

- npm/pnpm package is added with `@NpmPackage` or directly into [filename]`package.json`
- CSS or JavaScript is added with `@CssImport`, `@JsModule` or `@JavaScript`
- Vaadin add-on with front-end customisations is added

> - no `com.vaadin:vaadin-dev-bundle` dependency added to the project

Author: Project's theme is handled apart from bundle files, so it shouldn't generate a new bundle. Adding assets to the theme.json doesn't generate a new bundle, however it is accompanied usually with adding a new npm package, which per se triggers it.
What is missing here is "adding a packaged theme as a JAR dependency". Good point!

Custom theme files are served separately from the front-end bundle, i.e. they can be modified on the fly and the browser live-reload would trigger refresh immediately, no re-compilation is needed.
When customising a theme for the application in this mode, the following points should be taken into account:

> - Vaadin component's styles are set up in [filename]`styles.css` (or in the imported style sheet file located in the theme directory) with the `::part()` selector.

Reviewer 2: ```suggestion
- Vaadin component styles are set up in [filename]`styles.css` (or in the imported style sheet file located in the theme directory) with the `::part()` selector.
```

> - Use the `assets` block in the [filename]`theme.json` file to specify the external assets, like fonts, images or style sheets, to be added into the "application" development bundle, so that they can be visible and used in the application.

Reviewer 2: ```suggestion
- Use the `assets` block in the [filename]`theme.json` file to specify the external assets, like fonts, images or style sheets, to be added into the "application" development bundle, so that they can be used in the application.
```

Whereas, <<{articles}/production#,Production mode>> tries to send as little data as possible to the end user's browser. It optimizes the performance wherever possible. In this mode, no development tools are used, and debug information is disabled.

== Development Mode Methods

When using development mode, you have two choices or methods of how to utilize this mode:

- Use a Pre-Compiled Front-End Bundle; or
- Run with a Front-End Development Server.

The difference between these methods is in how the front-end files are handled.

When using a pre-compiled front-end bundle, the front-end files are collected and compiled into a bundle and are served to the browser similar to when in production mode. The difference is that debugging features are enabled.

When running with the front-end development server, JavaScript, CSS and other front-end files are served by it and given to the browser individually and as requested -- instead of the Java server of the application on which it's running.

[role="since:com.vaadin:vaadin@V24"]
== Pre-Compiled Front-End Bundle For Faster Start-Up

> The build and start-up time can be improved significantly if no front-end installation or compilation is made. This means not installing the tools Node.js, npm/pnpm and not downloading npm packages, nor running the Vite development server.

Reviewer 2: ```suggestion
The build and start-up time can be improved significantly if no front-end installation or compilation is made.
This means not installing the tools Node.js, npm/pnpm and not downloading npm packages, nor running the Vite development server.
```

Reviewer 2: Actually I'm not certain if we still use the one sentence one line.

If your project uses only the standard Vaadin Flow components or third-party Vaadin add-ons (e.g., from Vaadin Directory) without client-code, Vaadin skips front-end compilation and uses the default themes and front-end files shipped with the Vaadin platform.

> Adding any custom JavaScript/TypeScript or add-ons with front-end customizations to the project will trigger the front-end re-compilation with Vite during the next start-up of the application. Any new or missing front-end packages are downloaded using npm/pnpm before building. Vaadin does this automatically and is able to spot the front-end customizations.

Reviewer 2: ```suggestion
Adding any custom JavaScript/TypeScript or add-ons with front-end customizations to the project triggers the front-end re-compilation using Vite during the next start-up of the application.
Any new or missing front-end packages are downloaded using npm/pnpm before building.
Vaadin does this automatically and is able to spot the front-end customizations.
```
